mlir/tools/mlir-miopen-driver/mlir-miopen-driver.cpp

```diff
@@ -446,24 +459,35 @@ createConvertTensor(ModuleOp &module, OpBuilder &builder,
   // Emit loop body in the innermost loop.
   auto innermostLoopBuilder =
       OpBuilder::atBlockBegin(loopOpVector[loopOpVector.size() - 1].getBody());
-  auto dataType = originalMemRefType.getElementType();
+  auto convertedDataType = convertedMemRefType.getElementType();
   auto arguments = convertResultFuncOp.getArguments();
   mlir::Value sourceMemRef = arguments[0];
   mlir::Value convertedMemRef = arguments[1];
   mlir::Value sourceValue = innermostLoopBuilder.create<LoadOp>(
-      builder.getUnknownLoc(), dataType, sourceMemRef, loopIVVector);
+      builder.getUnknownLoc(), originalDataType, sourceMemRef, loopIVVector);
   mlir::Value convertedValue;
-  if (dataType == builder.getF16Type()) {
-    // Emit fpext.
-    convertedValue = innermostLoopBuilder.create<FPExtOp>(
-        builder.getUnknownLoc(), sourceValue, builder.getF32Type());
-  } else if (dataType == builder.getIntegerType(16)) {
-    // Treat I16 as BF16.
-    // TBD: Implement proper conversion logic. Force cast for now.
-    convertedValue = innermostLoopBuilder.create<SIToFPOp>(
-        builder.getUnknownLoc(), sourceValue, builder.getF32Type());
-  } else {
+  // convert to or from f16
+  if (originalDataType == convertedDataType) {
     convertedValue = sourceValue;
+  } else {
+    auto f16Type = builder.getF16Type();
+    auto f32Type = builder.getF32Type();
+    if ((originalDataType == f16Type && convertedDataType == f32Type) ||
+        (originalDataType == f32Type && convertedDataType == f16Type)) {
+      if (originalDataType == f16Type)
```

Comment:
the outer condition and the inner condition seems a bit redundant.
could we improve the branch condition here so we check the bit width of the types, instead of fixating on `f16` and `f32`?